experiments/kompanion/cmd/summary/summary.go

```diff
@@ -0,0 +1,398 @@
+	var resources []schema.GroupVersionResource
+	resources, err = utils.GetResources(discoveryClient, resources)
+	if err != nil {
+		return fmt.Errorf("error fetching resources: %w", err)
+	}
+
+	namespaces, err := clientset.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("error fetching namespaces: %w", err)
+	}
+
+	// create the work log for go routine workers to use
+	q := &taskQueue{}
+
+	summary := &Result{}
+
+	// Parallize across resources, unless we are scoped to a few namespaces
+	// The thought is that if users target a particular namespace (or a few), they may not have cluster-wide permission.
+	perNamespace := len(opts.targetNamespaces) > 0
+	if perNamespace {
+		for _, ns := range namespaces.Items {
+			if shouldExclude(ns.Name, opts.ignoreNamespaces, opts.targetNamespaces) {
+				continue
+			}
+
+			q.AddTask(&dumpResourcesTask{
+				Summary:       summary,
+				Namespace:     ns.Name,
+				Resources:     resources,
+				DynamicClient: dynamicClient,
+			})
+		}
+	} else {
+		for _, resource := range resources {
+			q.AddTask(&dumpResourcesTask{
+				Summary:       summary,
+				Resources:     []schema.GroupVersionResource{resource},
+				DynamicClient: dynamicClient,
+			})
+		}
+	}
+
+	log.Printf("Starting worker threads to process Config Connector objects")
+	var wg sync.WaitGroup
+
+	var errs []error
+	var errsMutex sync.Mutex
+
+	for i := 0; i < opts.workerRountines; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			for {
+				task := q.GetWork()
+				if task == nil {
+					// no work
+					return
+				}
+
+				//log.Printf("Starting work on %v", task)
+				if err := task.Run(ctx); err != nil {
+					errsMutex.Lock()
+					errs = append(errs, err)
+					errsMutex.Unlock()
+				}
+			}
+		}()
+	}
+
+	log.Printf("Dumping Config Connector summaries")
+	wg.Wait()
+
+	summary.Print()
+
+	return nil
+}
+
+func getStatus(r unstructured.Unstructured, gvr schema.GroupVersionResource) (string, error) {
+	id := types.NamespacedName{
+		Namespace: r.GetNamespace(),
+		Name:      r.GetName(),
+	}
+	slices, present, err := unstructured.NestedSlice(r.Object, "status", "conditions")
+	if !present {
+		return "no status", nil
+	}
+	if err != nil {
+		return "error", fmt.Errorf("error retrieving .status.healthy from resource %s: %w", id, err)
+	}
+	if len(slices) < 1 {
+		return "no status", nil
+	} else if len(slices) > 1 {
+		return "unknown", nil
+	}
+	element := slices[0]
+	condition, ok := element.(map[string]interface{})
+	if !ok {
+		return "unknown", fmt.Errorf("resource %s of type gvr %s failed to cast %T to condition", id, gvr, element)
+	}
+	val, present, err := unstructured.NestedString(condition, "status")
+	if !present {
+		return "unknown", nil
+	}
+	if err != nil {
+		return "unknown", fmt.Errorf("error retrieving .status.healthy from resource %s: %w", id, err)
+	}
+	return val, nil
+}
+
+func shouldExclude(name string, excludes []string, includes []string) bool {
```

Comment:
hehe we can probably factor some of these out but non blocking

Reply:
+1